### PR TITLE
Journalist can login to create an article

### PR DESCRIPTION
**Pull request by KyunginNa**
[Journalist can log in](https://www.pivotaltracker.com/story/show/176055862)
[Journalist can create article](https://www.pivotaltracker.com/story/show/176040561)
### User Story
```
As a journalist,
In order to write articles
I would like to be able to log in
```
```
As a journalist,
In order to write articles
I would like to be able to have an article saved in the API database
```
## Changes proposed in this pull request:
* Adding User model with devise_token_auth
* Creating a test and factory bot for User model
* Adding role to User model using enum 
* Adding article model and controller
* Creating a test and factory bot for Article model
* Creating `create` action in article controller with authentication and authorization
* Creating a test to check if only journalist can create an article
* Adding a route to create articles
## What we have learned working on this feature:
* The importance of prioritizing client requirements
* How to implement role-based authorization using enum
## Packages/Gems added
* devise_token_auth

### CODE REVIEW

File: app/models/article.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Article < ApplicationRecord
+  validates_presence_of :title, :content, :author
```
Comment (emtalen) [inline app/models/article.rb]: this has to belong to journalist

File: app/controllers/api/articles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+class Api::ArticlesController < ApplicationController
+  before_action :authenticate_user!, only: [:create]
+  before_action :is_user_journalist?, only: [:create]
+
+  def create
+    article = Article.create(article_params)
```
Comment (emtalen) [inline app/controllers/api/articles_controller.rb]: you want to set this to the current user straight away

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class User < ActiveRecord::Base
+  extend Devise::Models
+
+  after_initialize :set_default_role, if: :new_record?
+
+  enum role: [:visitor, :registered_user, :journalist]
```
Comment (emtalen) [inline app/models/user.rb]: remove visitor, as we spoke about

File: db/migrate/20201210163148_create_articles.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class CreateArticles < ActiveRecord::Migration[6.0]
+  def change
+    create_table :articles do |t|
+      t.string :title
+      t.text :sub_title
+      t.string :author
```
Comment (emtalen) [inline db/migrate/20201210163148_create_articles.rb:6]: you want this as a foreign key, not a string

File: db/migrate/20201210164453_add_role_to_users.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUsers < ActiveRecord::Migration[6.0]
+  def change
+    add_column :users, :role, :integer
```
Comment (emtalen) [inline db/migrate/20201210164453_add_role_to_users.rb:3]: you can set default here and it will by default be a registrered user

**Review by emtalen — CHANGES_REQUESTED** [review]

File: spec/models/article_spec.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    it { is_expected.to have_db_column :sub_title }
+    it { is_expected.to have_db_column :content}
+  end
+  describe 'Validations' do
+    it { is_expected.to validate_presence_of :title }
+    it { is_expected.to validate_presence_of :content}
+  end
+  describe 'Factory' do
+    it 'should have valid Factory' do
+      expect(FactoryBot.create(:article)).to be_valid
```
Comment (tochman) [inline spec/models/article_spec.rb]: You can omit `FactoryBot` if you have setup your `rails_helper` properly

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+RSpec.describe User, type: :model do
+  it "is expected to have valid Factory" do
+    expect(create(:user)).to be_valid
```
Comment (tochman) [inline spec/models/user_spec.rb:3]: You also have two other factories. Are you resting these?

**Review by tochman — CHANGES_REQUESTED** [review]
I see that you are not using Pundit for the authorization of the user roles.

File: spec/models/user_spec.rb (hunk 7 above)
Comment (KyunginNa) [reply spec/models/user_spec.rb:3]: We added tests for :registered_user and :journalist too

File: spec/models/article_spec.rb (hunk 6 above)
Comment (KyunginNa) [reply spec/models/article_spec.rb]: Deleted it and checked the test goes green.

**Review by emtalen — APPROVED** [review]

**Pull request merged**